tests/test_cli_compile.py

```diff
@@ -4044,3 +4045,78 @@ def test_second_order_requirements_relative_path_in_separate_dir(
             # via -r {output_path}
         """
     )
+
+
+@pytest.mark.parametrize(
+    "input_path_absolute", (True, False), ids=("absolute-input", "relative-input")
+)
+def test_url_constraints_are_not_treated_as_file_paths(
+    pip_conf,
+    make_package,
+    runner,
+    tmp_path,
+    monkeypatch,
+    input_path_absolute,
+):
+    """
+    Test normalization of ``-c`` constraints when the constraints are HTTPS URLs.
+    The constraints should be preserved verbatim.
+
+    This is a regression test for
+    https://github.com/jazzband/pip-tools/issues/2223
+    """
+    reqs_in = tmp_path / "requirements.in"
+    constraints_url = "https://example.com/files/common_constraints.txt"
```

Comment:
I'd swap these so that `reqs_in` initialization and use are close to each other:
```suggestion
    constraints_url = "https://example.com/files/common_constraints.txt"

    reqs_in = tmp_path / "requirements.in"
```